rust/transaction-filter/src/filters/user_transaction.rs

```diff
@@ -0,0 +1,160 @@
+            return Err(Error::msg("At least one of sender or payload must be set").into());
+        };
+        self.payload.is_valid()?;
+        Ok(())
+    }
+
+    #[inline]
+    fn is_allowed(&self, txn: &Transaction) -> bool {
+        let user_request = if let Some(TxnData::User(u)) = txn.txn_data.as_ref() {
+            if let Some(user_request) = u.request.as_ref() {
+                user_request
+            } else {
+                return false;
+            }
+        } else {
+            return false;
+        };
+
+        if let Some(sender_filter) = &self.sender {
+            if &user_request.sender != sender_filter {
+                return false;
+            }
+        }
+
+        if let Some(payload_filter) = &self.payload {
+            // Get the entry_function_payload from both UserPayload and MultisigPayload
+            let entry_function_payload = user_request
+                .payload
+                .as_ref()
+                .and_then(get_entry_function_payload_from_transaction_payload);
+            if let Some(payload) = entry_function_payload {
+                // Here we have an actual EntryFunctionPayload
+                if !payload_filter.is_allowed(payload) {
+                    return false;
+                }
+            }
+        }
+
+        true
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
+#[serde(deny_unknown_fields)]
+#[derive(derive_builder::Builder)]
+#[builder(setter(into, strip_option), default)]
+pub struct EntryFunctionFilter {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub address: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub module: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub function: Option<String>,
+}
+
+impl Filterable<EntryFunctionId> for EntryFunctionFilter {
+    #[inline]
+    fn validate_state(&self) -> Result<(), FilterError> {
+        if self.address.is_none() && self.module.is_none() && self.function.is_none() {
+            return Err(anyhow!("At least one of address, name or function must be set").into());
+        };
+        Ok(())
+    }
+
+    #[inline]
+    fn is_allowed(&self, module_id: &EntryFunctionId) -> bool {
+        if !self.module.is_allowed(&module_id.name) {
+            return false;
+        }
+
+        if self.address.is_some() || self.function.is_some() {
+            if let Some(module) = &module_id.module.as_ref() {
+                if !(self.address.is_allowed(&module.address)
+                    && self.function.is_allowed(&module.name))
+                {
+                    return false;
+                }
+            } else {
+                return false;
+            }
+        }
+
+        true
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
+#[serde(deny_unknown_fields)]
+#[derive(derive_builder::Builder)]
+#[builder(setter(strip_option), default)]
```

Comment:
I figure this shouldn't be `default`, because there is only one field so you must set it otherwise it'll always be invalid.

Reply:
It's required for the builder derive to allow partial completions- otherwise if you only set one field, and even though the overall builder is valid, it yells at you for fields being unset

I think there is some improvements we can make with it, I tried not to get too deep into it to avoid this PRs continued growth hahaha